### PR TITLE
Osher/mgmt 611

### PR DESCRIPTION
**Pull request by oshercc**

### CODE REVIEW

File: internal/bminventory/inventory.go [hunk 1]
```diff
@@ -607,6 +608,107 @@ func (b *bareMetalInventory) EnableHost(ctx context.Context, params inventory.En
...
+func (b *bareMetalInventory) createKubeconfigJob(ctx context.Context, cluster *models.Cluster) error {
+	id := cluster.ID
+	cfg, err := installcfg.GetInstallConfig(cluster)
+	if err != nil {
+		logrus.WithError(err).Errorf("failed to get install config for cluster %s", id)
+		return err
+	}
+	fmt.Println("Install config: \n", string(cfg))
+
+	if err := b.kube.Create(ctx, &batch.Job{
```
Comment (ronniel1) [inline internal/bminventory/inventory.go:621]: Why run it as a job and not just run a container using a goroutine?

File: internal/bminventory/inventory.go [hunk 2]
```diff
@@ -40,16 +40,17 @@ const (
 const (
 	ResourceKindHost    = "host"
 	ResourceKindCluster = "cluster"
-)
+	)
```
Comment (filanov) [inline internal/bminventory/inventory.go:43]: add file watcher to you goland

File: internal/bminventory/inventory.go [hunk 3]
```diff
@@ -40,16 +40,17 @@ const (
...
-	ImageBuilderCmd string `envconfig:"IMAGE_BUILDER_CMD" default:"echo hello"`
-	AgentDockerImg  string `envconfig:"AGENT_DOCKER_IMAGE" default:"quay.io/oamizur/introspector:latest"`
-	InventoryURL    string `envconfig:"INVENTORY_URL" default:"10.35.59.36"`
-	InventoryPort   string `envconfig:"INVENTORY_PORT" default:"30485"`
-	S3EndpointURL   string `envconfig:"S3_ENDPOINT_URL" default:"http://10.35.59.36:30925"`
-	S3Bucket        string `envconfig:"S3_BUCKET" default:"test"`
+	ImageBuilder    	string `envconfig:"IMAGE_BUILDER" default:"quay.io/oscohen/installer-image-build"`
+	ImageBuilderCmd 	string `envconfig:"IMAGE_BUILDER_CMD" default:"echo hello"`
+	AgentDockerImg  	string `envconfig:"AGENT_DOCKER_IMAGE" default:"quay.io/oamizur/introspector:latest"`
+	KubeconfigGenerator	string `envconfig:"IMAGE_BUILDER" default:"quay.io/oscohen/ignition-manifests-and-kubeconfig-generate"`
```
Comment (filanov) [inline internal/bminventory/inventory.go:49]: you have the same env for two different parameters `IMAGE_BUILDER`

File: internal/bminventory/inventory.go (hunk 1 above)
Comment (filanov) [reply internal/bminventory/inventory.go:626]: easy to handle retry and your image don't need to know how to run docker no?

File: internal/bminventory/inventory.go [hunk 4]
```diff
@@ -612,6 +613,106 @@ func (b *bareMetalInventory) EnableHost(ctx context.Context, params inventory.En
...
+	}
+	fmt.Println("Install config: \n", string(cfg))
+
+	if err := b.kube.Create(ctx, &batch.Job{
+		TypeMeta: meta.TypeMeta{
+			Kind:       "Job",
+			APIVersion: "batch/v1",
+		},
+		ObjectMeta: meta.ObjectMeta{
+			Name:      fmt.Sprintf("generate-kubeconfig-%s", id),
```
Comment (filanov) [inline internal/bminventory/inventory.go:632]: `generate-kubeconfig` appear in multiple places, make a const or function.

File: internal/bminventory/inventory.go [hunk 5]
```diff
@@ -612,6 +613,106 @@ func (b *bareMetalInventory) EnableHost(ctx context.Context, params inventory.En
...
+					Namespace: "default",
+				},
+				Spec: core.PodSpec{
+					Containers: []core.Container{
+						{
+							Name:            "generate-kubeconfig",
+							Image:           b.Config.KubeconfigGenerator,
+							Command:         b.imageBuildCmd,
+							ImagePullPolicy: "IfNotPresent",
+							Env: []core.EnvVar{
```
Comment (filanov) [inline internal/bminventory/inventory.go:649]: add a TODO adding s3 credentials.

File: internal/bminventory/inventory.go [hunk 6]
```diff
@@ -612,6 +613,106 @@ func (b *bareMetalInventory) EnableHost(ctx context.Context, params inventory.En
...
+	return nil
+}
+
 func (b *bareMetalInventory) DownloadClusterKubeconfig(ctx context.Context, params inventory.DownloadClusterKubeconfigParams) middleware.Responder {
-	return inventory.NewDownloadClusterKubeconfigNotFound()
+	var cluster models.Cluster
+
+	logrus.Infof("prepare and download kubeConfig for cluster %s", params.ClusterID)
+
+	tx := b.db.Begin()
```
Comment (filanov) [inline internal/bminventory/inventory.go:688]: why do you need a transaction? you are not changing anything in the DB right?

File: internal/bminventory/inventory.go [hunk 7]
```diff
@@ -612,6 +613,106 @@ func (b *bareMetalInventory) EnableHost(ctx context.Context, params inventory.En
...
+	}
+
+	if err := b.createKubeconfigJob(ctx, &cluster); err != nil {
+		logrus.WithError(err).Error("failed to create image job")
+		return inventory.NewDownloadClusterKubeconfigInternalServerError()
+	}
+
+	filesUrl := fmt.Sprintf("%s/%s/%s", b.S3EndpointURL, b.S3Bucket,
+		fmt.Sprintf("%s/kubeconfig", params.ClusterID))
+	logrus.Info("Image URL: ", filesUrl)
```
Comment (filanov) [inline internal/bminventory/inventory.go:711]: you don't have an image there.

File: internal/bminventory/inventory.go [hunk 8]
```diff
@@ -612,6 +613,106 @@ func (b *bareMetalInventory) EnableHost(ctx context.Context, params inventory.En
...
+	resp, err := http.Get(filesUrl)
+	if err != nil {
+		logrus.WithError(err).Error("failed to get image")
```
Comment (filanov) [inline internal/bminventory/inventory.go:714]: again not an image.

File: internal/bminventory/inventory.go [hunk 9]
```diff
@@ -23,6 +23,7 @@ import (
 )
 
 const baseHref = "/api/bm-inventory/v1"
+const generateKubeconfig = "generate-kubeconfig"
```
Comment (filanov) [inline internal/bminventory/inventory.go:26]: it's not clear what it used for maybe `kubeconfigPrefix` or something like that?

File: internal/bminventory/inventory.go [hunk 10]
```diff
@@ -43,13 +44,14 @@ const (
...
-	ImageBuilderCmd string `envconfig:"IMAGE_BUILDER_CMD" default:"echo hello"`
-	AgentDockerImg  string `envconfig:"AGENT_DOCKER_IMAGE" default:"quay.io/oamizur/introspector:latest"`
-	InventoryURL    string `envconfig:"INVENTORY_URL" default:"10.35.59.36"`
-	InventoryPort   string `envconfig:"INVENTORY_PORT" default:"30485"`
-	S3EndpointURL   string `envconfig:"S3_ENDPOINT_URL" default:"http://10.35.59.36:30925"`
-	S3Bucket        string `envconfig:"S3_BUCKET" default:"test"`
+	ImageBuilder    	string `envconfig:"IMAGE_BUILDER" default:"quay.io/oscohen/installer-image-build"`
+	ImageBuilderCmd 	string `envconfig:"IMAGE_BUILDER_CMD" default:"echo hello"`
+	AgentDockerImg  	string `envconfig:"AGENT_DOCKER_IMAGE" default:"quay.io/oamizur/introspector:latest"`
+	KubeconfigGenerator	string `envconfig:"KUBECONFIG_GENERATE" default:"quay.io/oscohen/ignition-manifests-and-kubeconfig-generate"`
```
Comment (filanov) [inline internal/bminventory/inventory.go:50]: Probably should be `KUBECONFIG_GENERATE_IMAGE`

File: internal/bminventory/inventory.go [hunk 11]
```diff
@@ -612,6 +615,95 @@ func (b *bareMetalInventory) EnableHost(ctx context.Context, params inventory.En
...
+
+func (b *bareMetalInventory) createKubeconfigJob(ctx context.Context, cluster *models.Cluster) error {
+	id := cluster.ID
+	cfg, err := installcfg.GetInstallConfig(cluster)
+	if err != nil {
+		logrus.WithError(err).Errorf("failed to get install config for cluster %s", id)
+		return err
+	}
+
+	fmt.Println("Install config: \n", string(cfg))
```
Comment (filanov) [inline internal/bminventory/inventory.go:627]: please remove or make it a log. 
if you are making it a log then change the content because it's not clear.

File: internal/bminventory/inventory.go [hunk 12]
```diff
@@ -612,6 +615,95 @@ func (b *bareMetalInventory) EnableHost(ctx context.Context, params inventory.En
...
+		return err
+	}
+	return nil
+}
+
 func (b *bareMetalInventory) DownloadClusterKubeconfig(ctx context.Context, params inventory.DownloadClusterKubeconfigParams) middleware.Responder {
-	return inventory.NewDownloadClusterKubeconfigNotFound()
+	var cluster models.Cluster
+
+	logrus.Infof("prepare and download kubeConfig for cluster %s", params.ClusterID)
```
Comment (filanov) [inline internal/bminventory/inventory.go:689]: prepare -> Prepare
maybe Prepare and download clusters <%s> kube-config? (not sure about the grammar)

File: internal/bminventory/inventory.go [hunk 13]
```diff
@@ -612,6 +615,95 @@ func (b *bareMetalInventory) EnableHost(ctx context.Context, params inventory.En
...
+
+	if err := b.db.Preload("Hosts").First(&cluster, "id = ?", params.ClusterID).Error; err != nil {
```
Comment (filanov) [inline internal/bminventory/inventory.go:691]: please add error log

File: internal/bminventory/inventory.go [hunk 14]
```diff
@@ -612,6 +615,95 @@ func (b *bareMetalInventory) EnableHost(ctx context.Context, params inventory.En
...
+		logrus.WithError(err).Error("failed to create image job")
+		return inventory.NewDownloadClusterKubeconfigInternalServerError()
+	}
+
+	filesUrl := fmt.Sprintf("%s/%s/%s", b.S3EndpointURL, b.S3Bucket,
+		fmt.Sprintf("%s/kubeconfig", params.ClusterID))
+	logrus.Info("File URL: ", filesUrl)
+	resp, err := http.Get(filesUrl)
+	if err != nil {
+		logrus.WithError(err).Error("failed to get File")
```
Comment (filanov) [inline internal/bminventory/inventory.go:705]: "Failed to get clusters <%s> kube-config"

File: subsystem/cluster_test.go [hunk 15]
```diff
@@ -139,5 +142,18 @@ var _ = Describe("system-test cluster install", func() {
 		for _, host := range c.GetPayload().Hosts {
 			Expect(swag.StringValue(host.Status)).Should(Equal("installing"))
 		}
+
+		file, err := ioutil.TempFile("", "tmp")
+		if err != nil {
```
Comment (filanov) [inline subsystem/cluster_test.go:147]: maybe should be
```suggestion
		Expect(err).NotTo(HaveOccurred())
```
You can fix the other place that doing the same as well.

File: subsystem/cluster_test.go [hunk 16]
```diff
@@ -139,5 +142,18 @@ var _ = Describe("system-test cluster install", func() {
...
+			log.Fatal(err)
+		}
+		defer os.Remove(file.Name())
+
+		_, err = bmclient.Inventory.DownloadClusterKubeconfig(ctx, &inventory.DownloadClusterKubeconfigParams{ClusterID: clusterID}, file)
+
```
Comment (filanov) [inline subsystem/cluster_test.go:153]: please remove empty line the next line is related directly to this one.

File: swagger.yaml [hunk 17]
```diff
@@ -181,11 +181,11 @@ paths:
       summary: Download the kubeconfig file for the specified cluster
       operationId: DownloadClusterKubeconfig
       produces:
-        - text/x-yaml
+        - application/octet-stream
       parameters:
         - in: path
           name: clusterId
-          description: The ID of the cluster whose kubeconfig to download
+          description: The ID of the clu  ster whose kubeconfig to download
```
Comment (filanov) [inline swagger.yaml:188]: typo

File: README.md [hunk 18]
```diff
@@ -67,3 +67,25 @@ The Other deployment is a system deployment it contain all the components the se
 This option will deploy S3 service (scality), DB and will use the image generator to create the images in the deployed S3 and create relevant bucket in S3.
 
 `skipper make deploy-all`
+
+
+##  Linked repositories
```
Comment (filanov) [inline README.md:72]: please make it in a different PR

File: deploy/bm-inventory.yaml [hunk 19]
```diff
@@ -14,7 +14,7 @@ spec:
     spec:
       containers:
         - name: bm-inventory
-          image: REPLACE_IMAGE
+          image: quay.io/oscohen/bm-inventory:latest
```
Comment (filanov) [inline deploy/bm-inventory.yaml:17]: again no!

File: internal/bminventory/inventory.go [hunk 20]
```diff
@@ -612,6 +615,94 @@ func (b *bareMetalInventory) EnableHost(ctx context.Context, params inventory.En
...
+
+	logrus.Infof("Prepare and download kubeConfig from cluster %s", params.ClusterID)
+
+	if err := b.db.Preload("Hosts").First(&cluster, "id = ?", params.ClusterID).Error; err != nil {
+		logrus.WithError(err).Errorf("failed to get cluster %s", params.ClusterID)
+		return inventory.NewDownloadClusterKubeconfigNotFound()
+	}
+
+	if err := b.createKubeconfigJob(ctx, &cluster); err != nil {
+		logrus.WithError(err).Error("failed to create image job")
```
Comment (filanov) [inline internal/bminventory/inventory.go:695]: Capital

File: subsystem/cluster_test.go [hunk 21]
```diff
@@ -90,7 +93,7 @@ var _ = Describe("Cluster tests", func() {
 	})
 })
 
-var _ = Describe("system-test cluster install", func() {
+	var _ = Describe("system-test cluster install", func() {
```
Comment (filanov) [inline subsystem/cluster_test.go:96]: ?

File: internal/bminventory/inventory.go [hunk 22]
```diff
@@ -612,6 +615,94 @@ func (b *bareMetalInventory) EnableHost(ctx context.Context, params inventory.En
 	return inventory.NewEnableHostNoContent()
 }
 
+
+func (b *bareMetalInventory) createKubeconfigJob(ctx context.Context, cluster *models.Cluster) error {
+	id := cluster.ID
+	cfg, err := installcfg.GetInstallConfig(cluster)
+	if err != nil {
+		logrus.WithError(err).Errorf("failed to get install config for cluster %s", id)
```
Comment (filanov) [inline internal/bminventory/inventory.go:623]: Capital

File: internal/bminventory/inventory.go [hunk 23]
```diff
@@ -612,6 +615,94 @@ func (b *bareMetalInventory) EnableHost(ctx context.Context, params inventory.En
...
+	var cluster models.Cluster
+
+	logrus.Infof("Prepare and download kubeConfig from cluster %s", params.ClusterID)
+
+	if err := b.db.Preload("Hosts").First(&cluster, "id = ?", params.ClusterID).Error; err != nil {
+		logrus.WithError(err).Errorf("failed to get cluster %s", params.ClusterID)
+		return inventory.NewDownloadClusterKubeconfigNotFound()
+	}
+
+	if err := b.createKubeconfigJob(ctx, &cluster); err != nil {
```
Comment (filanov) [inline internal/bminventory/inventory.go:694]: you created a job but you don't monitor it
what if it will take a long time and your get will fail?

File: internal/bminventory/inventory.go [hunk 24]
```diff
@@ -102,19 +104,19 @@ func buildHrefURI(base, id string) *strfmt.URI {
 	return strToURI(fmt.Sprintf("%s/%ss/%s", baseHref, base, id))
 }
 
-func (b *bareMetalInventory) monitorImageBuild(ctx context.Context, id string) error {
+func (b *bareMetalInventory) monitorImageBuild(ctx context.Context, jobName string) error {
```
Comment (filanov) [inline internal/bminventory/inventory.go:107]: it doesn't monitor image build anymore.
```suggestion
func (b *bareMetalInventory) monitorJobProgress(ctx context.Context, jobName string) error {
```

File: subsystem/cluster_test.go [hunk 25]
```diff
@@ -139,5 +141,16 @@ var _ = Describe("system-test cluster install", func() {
 		for _, host := range c.GetPayload().Hosts {
 			Expect(swag.StringValue(host.Status)).Should(Equal("installing"))
 		}
+
+		file, err := ioutil.TempFile("", "tmp")
```
Comment (filanov) [inline subsystem/cluster_test.go:145]: please add a bad case where cluster id doesn't exists

File: subsystem/cluster_test.go [hunk 26]
```diff
@@ -2,13 +2,17 @@ package subsystem
...
+	"os"
 
 	"github.com/filanov/bm-inventory/client/inventory"
 	"github.com/filanov/bm-inventory/models"
 	"github.com/go-openapi/strfmt"
 	"github.com/go-openapi/swag"
+	"github.com/google/uuid"
 	. "github.com/onsi/ginkgo"
 	. "github.com/onsi/gomega"
+
```
Comment (filanov) [inline subsystem/cluster_test.go:15]: did you enabled file watcher?

File: internal/bminventory/inventory.go [hunk 27]
```diff
@@ -102,19 +104,19 @@ func buildHrefURI(base, id string) *strfmt.URI {
 	return strToURI(fmt.Sprintf("%s/%ss/%s", baseHref, base, id))
 }
 
-func (b *bareMetalInventory) monitorImageBuild(ctx context.Context, id string) error {
+func (b *bareMetalInventory) monitorJobBuild(ctx context.Context, jobName string) error {
```
Comment (filanov) [inline internal/bminventory/inventory.go:107]: why build? what is build here?

File: subsystem/cluster_test.go (hunk 26 above)
Comment (oshercc) [reply subsystem/cluster_test.go:15]: yes
